lib/src/query_snapshot_stream_manager.dart

```diff
@@ -81,25 +87,118 @@ class QuerySnapshotStreamManager {
     return streamController;
   }
 
-  Future<void> fireSnapshotUpdate(
-      FirebaseFirestore firestore, String path) async {
+  Future<void> fireSnapshotUpdate<T>(
+    FirebaseFirestore firestore,
+    String path, {
+    String? id,
+  }) async {
     if (!_streamCache.containsKey(firestore)) {
       // Normal. It happens if you try to fire updates before anyone has
       // subscribed to snapshots.
       return;
     }
     final exactPathCache = _streamCache[firestore]![path];
-    if (exactPathCache != null) {
+    if (exactPathCache != null && id != null) {
       for (final query in [...exactPathCache.keys]) {
-        await query.get().then(exactPathCache[query]!.add);
+        if (query is! FakeQueryWithParent<T>) {
+          continue;
+        }
+
+        final invalidCache = _cacheQuerySnapshot[query] != null &&
+            _cacheQuerySnapshot[query] is! QuerySnapshot<T>;
+        if (invalidCache) {
+          assert(invalidCache,
+              'querySnapshotPrior is not null or QuerySnapshot<T>. Got ${_cacheQuerySnapshot[query]}');
+          continue;
+        }
+        final querySnapshotPrior =
+            _cacheQuerySnapshot[query] as QuerySnapshot<T>?;
+
+        final querySnapshot = await query.get();
+        _cacheQuerySnapshot[query] = querySnapshot;
+        final _docsPrior = querySnapshotPrior?.docs ?? [];
+        final _docsCurrent = List.of(querySnapshot.docs);
+
+        final _docChange = _getDocumentChange<T>(
+          id: id,
+          docsPrior: _docsPrior,
+          docsCurrent: _docsCurrent,
+        );
+
+        final _documentsChange = <DocumentChange<T>>[];
+
+        if (_docChange != null) {
+          _documentsChange.add(_docChange);
+        }
+        final _querySnapshot = MockQuerySnapshot<T>(
+          _docsCurrent,
+          querySnapshot.metadata.isFromCache,
+          documentChanges: _documentsChange,
+        );
+        exactPathCache[query]?.add(_querySnapshot);
       }
     }
 
     // When a document is modified, fire an update on the parent collection.
     if (path.contains('/')) {
       final tokens = path.split('/');
       final parentPath = tokens.sublist(0, tokens.length - 1).join('/');
-      await fireSnapshotUpdate(firestore, parentPath);
+      final _id = id ?? tokens.last;
+      await fireSnapshotUpdate<T>(firestore, parentPath, id: _id);
     }
   }
+
+  /// Returns [DocumentChange] for doc [id] based on the change between [docsPrior] and [docsCurrent].
+  DocumentChange<T>? _getDocumentChange<T>({
+    required String id,
+    required List<QueryDocumentSnapshot<T>> docsPrior,
+    required List<QueryDocumentSnapshot<T>> docsCurrent,
+  }) {
+    final _docPriorIndex = docsPrior.indexWhere((element) {
+      return element.id == id;
+    });
+    QueryDocumentSnapshot<T>? _docPrior;
+    if (_docPriorIndex != -1) {
+      _docPrior = docsPrior[_docPriorIndex];
+    }
```

Comment:
Since you only need this on line 188, how about inlining it there like you did on line 180 for docsCurrent?

```dart
   else if (_docCurrentIndex == -1 && _docPriorIndex != -1) {
      /// Document is removed.
      return MockDocumentChange<T>(
        docsPrior[_docPriorIndex]!,
```